File: python-alert-service/repositories/alert_rule_repository.py
"""Repository for alert rule data access."""

import sqlite3
from datetime import datetime, timezone
from typing import Optional

from fastapi import Depends

from database import get_db_connection
from models.alert_rule import AlertRule, AlertRuleCreate, AlertRuleUpdate


class AlertRuleRepository:
    """Data access layer for alert rules using SQLite."""

    def __init__(self, db: sqlite3.Connection):
        self.db = db

    def get_all(self) -> list[AlertRule]:
        """Retrieve all alert rules."""
        cursor = self.db.execute(
            "SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at "
            "FROM alert_rules ORDER BY id"
        )
        rows = cursor.fetchall()
        return [self._row_to_rule(row) for row in rows]

    def get_by_id(self, rule_id: str) -> Optional[AlertRule]:
        """Retrieve an alert rule by ID."""
        cursor = self.db.execute(
            "SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at "
            "FROM alert_rules WHERE id = ?",
            (rule_id,),
        )
        row = cursor.fetchone()
        return self._row_to_rule(row) if row else None

    def get_active_rules_for_sensor(self, sensor_id: str) -> list[AlertRule]:
        """Retrieve all active rules for a given sensor."""
        cursor = self.db.execute(
            "SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at "
            "FROM alert_rules WHERE sensor_id = ? AND status = 'active' ORDER BY id",
            (sensor_id,),
        )
        rows = cursor.fetchall()
        return [self._row_to_rule(row) for row in rows]

    def create(self, rule: AlertRuleCreate) -> AlertRule:
        """Create a new alert rule."""
        cursor = self.db.execute(
            "SELECT MAX(CAST(SUBSTR(id, 6) AS INTEGER)) FROM alert_rules WHERE id LIKE 'rule-%'"
        )
        max_num = cursor.fetchone()[0] or 0
        new_id = f"rule-{max_num + 1:03d}"

        now = datetime.now(timezone.utc).isoformat()

        self.db.execute(
            """
            INSERT INTO alert_rules (id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                new_id,
                rule.sensor_id,
                rule.metric,
                rule.operator.value,
                rule.threshold,
                rule.name,
                rule.status.value,
                now,
                now,
            ),
        )
        self.db.commit()

        return self.get_by_id(new_id)

    def update(self, rule_id: str, updates: AlertRuleUpdate) -> Optional[AlertRule]:
        """Update an existing alert rule."""
        existing = self.get_by_id(rule_id)
        if not existing:
            return None

        update_fields = []
        values = []

        update_data = updates.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            if value is not None:
                if field in ("operator", "status"):
                    value = value.value if hasattr(value, "value") else value
                update_fields.append(f"{field} = ?")
                values.append(value)

        if not update_fields:
            return existing

        now = datetime.now(timezone.utc).isoformat()
        update_fields.append("updated_at = ?")
        values.extend([now, rule_id])

        query = f"UPDATE alert_rules SET {', '.join(update_fields)} WHERE id = ?"
        self.db.execute(query, values)
        self.db.commit()

        return self.get_by_id(rule_id)

    def delete(self, rule_id: str) -> bool:
        """Delete an alert rule by ID."""
        cursor = self.db.execute("DELETE FROM alert_rules WHERE id = ?", (rule_id,))
        self.db.commit()
        return cursor.rowcount > 0

    def _row_to_rule(self, row: sqlite3.Row) -> AlertRule:
        """Convert a database row to an AlertRule model."""
        return AlertRule(
            id=row["id"],
            sensor_id=row["sensor_id"],
            metric=row["metric"],
            operator=row["operator"],
            threshold=row["threshold"],
            name=row["name"],
            status=row["status"],
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )


def get_alert_rule_repository(
    db: sqlite3.Connection = Depends(get_db_connection),
) -> AlertRuleRepository:
    """Dependency that provides an AlertRuleRepository instance."""
    return AlertRuleRepository(db)

File: go-alert-service/repositories/alert_rule_repository.go
package repositories

import (
	"database/sql"
	"fmt"

	"iot-alert-service/models"
)

// AlertRuleRepository defines the interface for alert rule data access.
type AlertRuleRepository interface {
	GetAll() ([]models.AlertRule, error)
	GetByID(id string) (*models.AlertRule, error)
	GetActiveRulesForSensor(sensorID string) ([]models.AlertRule, error)
	Create(rule *models.AlertRuleCreate) (*models.AlertRule, error)
	Update(id string, updates *models.AlertRuleUpdate) (*models.AlertRule, error)
	Delete(id string) error
}

// PostgresAlertRuleRepository implements AlertRuleRepository using Postgres.
type PostgresAlertRuleRepository struct {
	db *sql.DB
}

// NewAlertRuleRepository creates a new Postgres-backed alert rule repository.
func NewAlertRuleRepository(db *sql.DB) *PostgresAlertRuleRepository {
	return &PostgresAlertRuleRepository{db: db}
}

// GetAll retrieves all alert rules from the database.
func (r *PostgresAlertRuleRepository) GetAll() ([]models.AlertRule, error) {
	rows, err := r.db.Query(`
		SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at
		FROM alert_rules ORDER BY id
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var rules []models.AlertRule
	for rows.Next() {
		var rule models.AlertRule
		err := rows.Scan(&rule.ID, &rule.SensorID, &rule.Metric, &rule.Operator,
			&rule.Threshold, &rule.Name, &rule.Status, &rule.CreatedAt, &rule.UpdatedAt)
		if err != nil {
			return nil, err
		}
		rules = append(rules, rule)
	}

	if rules == nil {
		rules = []models.AlertRule{}
	}

	return rules, rows.Err()
}

// GetByID retrieves an alert rule by its ID.
func (r *PostgresAlertRuleRepository) GetByID(id string) (*models.AlertRule, error) {
	var rule models.AlertRule
	err := r.db.QueryRow(`
		SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at
		FROM alert_rules WHERE id = $1
	`, id).Scan(&rule.ID, &rule.SensorID, &rule.Metric, &rule.Operator,
		&rule.Threshold, &rule.Name, &rule.Status, &rule.CreatedAt, &rule.UpdatedAt)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &rule, nil
}

// GetActiveRulesForSensor retrieves all active rules for a given sensor.
func (r *PostgresAlertRuleRepository) GetActiveRulesForSensor(sensorID string) ([]models.AlertRule, error) {
	rows, err := r.db.Query(`
		SELECT id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at
		FROM alert_rules WHERE sensor_id = $1 AND status = 'active' ORDER BY id
	`, sensorID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var rules []models.AlertRule
	for rows.Next() {
		var rule models.AlertRule
		err := rows.Scan(&rule.ID, &rule.SensorID, &rule.Metric, &rule.Operator,
			&rule.Threshold, &rule.Name, &rule.Status, &rule.CreatedAt, &rule.UpdatedAt)
		if err != nil {
			return nil, err
		}
		rules = append(rules, rule)
	}

	if rules == nil {
		rules = []models.AlertRule{}
	}

	return rules, rows.Err()
}

// Create inserts a new alert rule into the database.
// Uses a transaction to ensure atomic ID generation and insertion,
// preventing duplicate IDs under concurrent access.
func (r *PostgresAlertRuleRepository) Create(rule *models.AlertRuleCreate) (*models.AlertRule, error) {
	if err := rule.Validate(); err != nil {
		return nil, err
	}

	tx, err := r.db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	var maxNum sql.NullInt64
	err = tx.QueryRow("SELECT MAX(CAST(SUBSTR(id, 6) AS INTEGER)) FROM alert_rules WHERE id LIKE 'rule-%'").Scan(&maxNum)
	if err != nil {
		return nil, err
	}

	nextNum := int64(1)
	if maxNum.Valid {
		nextNum = maxNum.Int64 + 1
	}
	newID := fmt.Sprintf("rule-%03d", nextNum)

	now := models.Now()

	_, err = tx.Exec(`
		INSERT INTO alert_rules (id, sensor_id, metric, operator, threshold, name, status, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
	`, newID, rule.SensorID, rule.Metric, rule.Operator, rule.Threshold, rule.Name, rule.Status, now, now)

	if err != nil {
		return nil, err
	}

	if err := tx.Commit(); err != nil {
		return nil, err
	}

	return r.GetByID(newID)
}

// Update modifies an existing alert rule.
func (r *PostgresAlertRuleRepository) Update(id string, updates *models.AlertRuleUpdate) (*models.AlertRule, error) {
	existing, err := r.GetByID(id)
	if err != nil {
		return nil, err
	}
	if existing == nil {
		return nil, nil
	}

	if err := updates.Validate(); err != nil {
		return nil, err
	}

	argIdx := 2
	query := "UPDATE alert_rules SET updated_at = $1"
	args := []interface{}{models.Now()}

	if updates.SensorID != nil {
		query += fmt.Sprintf(", sensor_id = $%d", argIdx)
		args = append(args, *updates.SensorID)
		argIdx++
	}
	if updates.Metric != nil {
		query += fmt.Sprintf(", metric = $%d", argIdx)
		args = append(args, *updates.Metric)
		argIdx++
	}
	if updates.Operator != nil {
		query += fmt.Sprintf(", operator = $%d", argIdx)
		args = append(args, *updates.Operator)
		argIdx++
	}
	if updates.Threshold != nil {
		query += fmt.Sprintf(", threshold = $%d", argIdx)
		args = append(args, *updates.Threshold)
		argIdx++
	}
	if updates.Name != nil {
		query += fmt.Sprintf(", name = $%d", argIdx)
		args = append(args, *updates.Name)
		argIdx++
	}
	if updates.Status != nil {
		query += fmt.Sprintf(", status = $%d", argIdx)
		args = append(args, *updates.Status)
		argIdx++
	}

	query += fmt.Sprintf(" WHERE id = $%d", argIdx)
	args = append(args, id)

	_, err = r.db.Exec(query, args...)
	if err != nil {
		return nil, err
	}

	return r.GetByID(id)
}

// Delete removes an alert rule from the database.
func (r *PostgresAlertRuleRepository) Delete(id string) error {
	result, err := r.db.Exec("DELETE FROM alert_rules WHERE id = $1", id)
	if err != nil {
		return err
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return err
	}

	if rowsAffected == 0 {
		return sql.ErrNoRows
	}

	return nil
}
